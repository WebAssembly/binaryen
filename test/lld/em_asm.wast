(module
 (type $0 (func (param i32) (result i32)))
 (type $1 (func (param i32 i32 i32) (result i32)))
 (type $2 (func (param i32 i32) (result i32)))
 (type $3 (func (result i32)))
 (type $4 (func))
 (import "env" "_Z24emscripten_asm_const_intIJEEiPKcDpT_" (func $_Z24emscripten_asm_const_intIJEEiPKcDpT_ (param i32) (result i32)))
 (import "env" "_Z24emscripten_asm_const_intIJiiEEiPKcDpT_" (func $_Z24emscripten_asm_const_intIJiiEEiPKcDpT_ (param i32 i32 i32) (result i32)))
 (import "env" "_Z24emscripten_asm_const_intIJiEEiPKcDpT_" (func $_Z24emscripten_asm_const_intIJiEEiPKcDpT_ (param i32 i32) (result i32)))
 (global $global$0 (mut i32) (i32.const 66192))
 (global $global$1 i32 (i32.const 66192))
 (global $global$2 i32 (i32.const 652))
 (table 1 1 funcref)
 (memory $0 2)
 (data (i32.const 568) "{ Module.print(\"Hello world\"); }\00{ return $0 + $1; }\00{ Module.print(\"Got \" + $0); }\00")
 (export "memory" (memory $0))
 (export "__wasm_call_ctors" (func $__wasm_call_ctors))
 (export "main" (func $main))
 (export "__heap_base" (global $global$1))
 (export "__data_end" (global $global$2))
 (func $main (; 3 ;) (type $3) (result i32)
  (drop
   (call $_Z24emscripten_asm_const_intIJEEiPKcDpT_
    (i32.const 568)
   )
  )
  (drop
   (call $_Z24emscripten_asm_const_intIJiEEiPKcDpT_
    (i32.const 621)
    (call $_Z24emscripten_asm_const_intIJiiEEiPKcDpT_
     (i32.const 601)
     (i32.const 13)
     (i32.const 27)
    )
   )
  )
  (i32.const 0)
 )
 (func $__wasm_call_ctors (; 4 ;) (type $4)
 )
 ;; custom section "linking", size 3
)

