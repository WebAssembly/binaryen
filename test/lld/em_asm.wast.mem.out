(module
 (type $0 (func (param i32 i32 i32) (result i32)))
 (type $1 (func))
 (type $2 (func (result i32)))
 (type $3 (func (param i32 i32 i32)))
 (type $4 (func (param i32 i32)))
 (type $5 (func (param i32) (result i32)))
 (type $6 (func (param i32 i32) (result i32)))
 (type $FUNCSIG$iiii (func (param i32 i32 i32) (result i32)))
 (import "env" "emscripten_asm_const_iii" (func $emscripten_asm_const_iii (param i32 i32 i32) (result i32)))
 (memory $0 2)
 (table $0 1 1 funcref)
 (global $global$0 (mut i32) (i32.const 16384))
 (global $global$1 i32 (i32.const 66192))
 (global $global$2 i32 (i32.const 652))
 (export "memory" (memory $0))
 (export "__wasm_call_ctors" (func $__wasm_call_ctors))
 (export "__heap_base" (global $global$1))
 (export "__data_end" (global $global$2))
 (export "main" (func $main))
 (export "stackSave" (func $stackSave))
 (export "stackAlloc" (func $stackAlloc))
 (export "stackRestore" (func $stackRestore))
 (export "__growWasmMemory" (func $__growWasmMemory))
 (func $__wasm_call_ctors (; 1 ;) (type $1)
  (nop)
 )
 (func $__original_main (; 2 ;) (type $2) (result i32)
  (local $0 i32)
  (local $1 i32)
  (global.set $global$0
   (local.tee $0
    (i32.sub
     (global.get $global$0)
     (i32.const 32)
    )
   )
  )
  (i32.store8 offset=24
   (local.get $0)
   (call $__em_asm_sig_builder::inner<>\20const\20__em_asm_sig_builder::__em_asm_sig<>\28\29)
  )
  (drop
   (call $emscripten_asm_const_iii
    (i32.const 0)
    (i32.add
     (local.get $0)
     (i32.const 24)
    )
    (i32.const 0)
   )
  )
  (call $__em_asm_sig_builder::inner<int\2c\20int>\20const\20__em_asm_sig_builder::__em_asm_sig<int\2c\20int>\28int\2c\20int\29
   (i32.add
    (local.get $0)
    (i32.const 24)
   )
   (i32.const 13)
   (i32.const 27)
  )
  (i64.store offset=16
   (local.get $0)
   (i64.const 115964117005)
  )
  (call $__em_asm_sig_builder::inner<int>\20const\20__em_asm_sig_builder::__em_asm_sig<int>\28int\29
   (i32.add
    (local.get $0)
    (i32.const 24)
   )
   (local.tee $1
    (call $emscripten_asm_const_iii
     (i32.const 1)
     (i32.add
      (local.get $0)
      (i32.const 24)
     )
     (i32.add
      (local.get $0)
      (i32.const 16)
     )
    )
   )
  )
  (i32.store
   (local.get $0)
   (local.get $1)
  )
  (drop
   (call $emscripten_asm_const_iii
    (i32.const 2)
    (i32.add
     (local.get $0)
     (i32.const 24)
    )
    (local.get $0)
   )
  )
  (global.set $global$0
   (i32.add
    (local.get $0)
    (i32.const 32)
   )
  )
  (i32.const 0)
 )
 (func $__em_asm_sig_builder::inner<>\20const\20__em_asm_sig_builder::__em_asm_sig<>\28\29 (; 3 ;) (type $2) (result i32)
  (i32.const 0)
 )
 (func $__em_asm_sig_builder::inner<int\2c\20int>\20const\20__em_asm_sig_builder::__em_asm_sig<int\2c\20int>\28int\2c\20int\29 (; 4 ;) (type $3) (param $0 i32) (param $1 i32) (param $2 i32)
  (local $3 i32)
  (global.set $global$0
   (local.tee $3
    (i32.sub
     (global.get $global$0)
     (i32.const 16)
    )
   )
  )
  (i32.store8 offset=13
   (local.get $3)
   (call $__em_asm_sig_builder::sig_char\28int\29
    (local.get $1)
   )
  )
  (local.set $2
   (call $__em_asm_sig_builder::sig_char\28int\29
    (local.get $2)
   )
  )
  (i32.store8
   (i32.add
    (local.get $0)
    (i32.const 2)
   )
   (i32.const 0)
  )
  (i32.store8 offset=14
   (local.get $3)
   (local.get $2)
  )
  (i32.store16 align=1
   (local.get $0)
   (i32.load16_u offset=13 align=1
    (local.get $3)
   )
  )
  (global.set $global$0
   (i32.add
    (local.get $3)
    (i32.const 16)
   )
  )
 )
 (func $__em_asm_sig_builder::inner<int>\20const\20__em_asm_sig_builder::__em_asm_sig<int>\28int\29 (; 5 ;) (type $4) (param $0 i32) (param $1 i32)
  (local $2 i32)
  (global.set $global$0
   (local.tee $2
    (i32.sub
     (global.get $global$0)
     (i32.const 16)
    )
   )
  )
  (local.set $1
   (call $__em_asm_sig_builder::sig_char\28int\29
    (local.get $1)
   )
  )
  (i32.store8 offset=15
   (local.get $2)
   (i32.const 0)
  )
  (i32.store8 offset=14
   (local.get $2)
   (local.get $1)
  )
  (i32.store16 align=1
   (local.get $0)
   (i32.load16_u offset=14
    (local.get $2)
   )
  )
  (global.set $global$0
   (i32.add
    (local.get $2)
    (i32.const 16)
   )
  )
 )
 (func $__em_asm_sig_builder::sig_char\28int\29 (; 6 ;) (type $5) (param $0 i32) (result i32)
  (i32.const 105)
 )
 (func $main (; 7 ;) (type $6) (param $0 i32) (param $1 i32) (result i32)
  (call $__original_main)
 )
 (func $stackSave (; 8 ;) (result i32)
  (global.get $global$0)
 )
 (func $stackAlloc (; 9 ;) (param $0 i32) (result i32)
  (local $1 i32)
  (global.set $global$0
   (local.tee $1
    (i32.and
     (i32.sub
      (global.get $global$0)
      (local.get $0)
     )
     (i32.const -16)
    )
   )
  )
  (local.get $1)
 )
 (func $stackRestore (; 10 ;) (param $0 i32)
  (global.set $global$0
   (local.get $0)
  )
 )
 (func $__growWasmMemory (; 11 ;) (param $newSize i32) (result i32)
  (grow_memory
   (local.get $newSize)
  )
 )
)
(;
--BEGIN METADATA --
{
  "asmConsts": {
    "2": ["{ Module.print(\"Got \" + $0); }", ["iii"], [""]],
    "0": ["{ Module.print(\"Hello world\"); }", ["iii"], [""]],
    "1": ["{ return $0 + $1; }", ["iii"], [""]]
  },
  "staticBump": 84,
  "tableSize": 1,
  "initializers": [
    "__wasm_call_ctors"
  ],
  "declares": [
  ],
  "externs": [
  ],
  "implementedFunctions": [
    "___wasm_call_ctors",
    "_main",
    "_stackSave",
    "_stackAlloc",
    "_stackRestore",
    "___growWasmMemory"
  ],
  "exports": [
    "memory",
    "__wasm_call_ctors",
    "__heap_base",
    "__data_end",
    "main",
    "stackSave",
    "stackAlloc",
    "stackRestore",
    "__growWasmMemory"
  ],
  "invokeFuncs": [
  ]
}
-- END METADATA --
;)
